### PR TITLE
refactor: select only 10 notifications form DB

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

when selecting notifications form DB we fetch all available notifications and filter the first 10 in logic 

### Solutions

use window functions to select only 10 notificaions form the DB

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/GetNotificationsUseCase.kt [hunk 1]
```diff
@@ -81,7 +80,6 @@ internal class GetNotificationsUseCaseImpl internal constructor(
                         list.filter { it !is LocalNotification.Conversation || it.messages.isNotEmpty() }
                     }
             }
-            .distinctUntilChanged()
```
Comment (borichellow) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/GetNotificationsUseCase.kt:84]: why you removed it?

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt [hunk 2]
```diff
@@ -292,12 +292,11 @@ internal class MessageDAOImpl internal constructor(
                 .associateBy { it.conversationId }
         }
 
-    override suspend fun getNotificationMessage(): Flow<List<NotificationMessageEntity>> =
+    override suspend fun getNotificationMessage(maxNumberOfMessagesPerConversation: Int): Flow<List<NotificationMessageEntity>> =
```
Comment (borichellow) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt:295]: what is the point of adding that parameter when it's not used?

**Review by Garzas — APPROVED** [review]

**Review by borichellow — APPROVED** [review]
Good idea! 
just a few comments.

**Review by mchenani — COMMENTED** [review]
With this solution we always get the last N messages and it will fix one issue, but if we want to show grouping notifications, like 3 last messages of all unread conversation, this won't work.
I suggest that you take a look on window queries, since SQL Delight supports them now.
an example:
```
WITH RankedMessages AS (
  SELECT
    *,
    ROW_NUMBER() OVER (PARTITION BY conversation_id ORDER BY message_timestamp DESC) AS message_rank
  FROM
    messages
)
SELECT
  *
FROM
  RankedMessages
WHERE
  message_rank <= 3;
```

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt (hunk 2 above)
Comment (MohamadJaara) [reply persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAOImpl.kt:295]: i wanted to use it but i caused error message in from sql delight psi, i will remove it

**Comment by mchenani** [comment]
> With this solution we always get the last N messages and it will fix one issue, but if we want to show grouping notifications, like 3 last messages of all unread conversation, this won't work. I suggest that you take a look on window queries, since SQL Delight supports them now. an example:
> 
> ```
> WITH RankedMessages AS (
>   SELECT
>     *,
>     ROW_NUMBER() OVER (PARTITION BY conversation_id ORDER BY message_timestamp DESC) AS message_rank
>   FROM
>     messages
> )
> SELECT
>   *
> FROM
>   RankedMessages
> WHERE
>   message_rank <= 3;
> ```

I retract my comment, since it's already covered. I missed that part in the code. Thanks

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/GetNotificationsUseCase.kt (hunk 1 above)
Comment (MohamadJaara) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/GetNotificationsUseCase.kt:84]: it is not needed in this context

**Pull request merged**